From 01a7d9ed511549631be7d0fa93cea8a44457260e Mon Sep 17 00:00:00 2001
From: Vishal Sagar <vishal.sagar@xilinx.com>
Date: Wed, 14 Oct 2020 04:51:06 -0700
Subject: [PATCH 1711/1851] v4l: xilinx: sdirxss: Fix colorimetry dectection
 for HD mode

commit 95f4bbb110db2f699b19891ecc0bcc6c40d357e1 from
https://github.com/Xilinx/linux-xlnx.git

For HD mode, bit 4 and 7 of byte 3 (bit 23 and 20) of ST352
payload determine the colorimetry per SMPTE 292-1:2018 Sec 9.5.

The same thing is observed with Phabrix Qx for 3GB mode with payload 0x8A.

Signed-off-by: Vishal Sagar <vishal.sagar@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/media/platform/xilinx/xilinx-sdirxss.c | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/drivers/media/platform/xilinx/xilinx-sdirxss.c b/drivers/media/platform/xilinx/xilinx-sdirxss.c
index 8b7c335bea13..75d2217ca030 100644
--- a/drivers/media/platform/xilinx/xilinx-sdirxss.c
+++ b/drivers/media/platform/xilinx/xilinx-sdirxss.c
@@ -16,6 +16,7 @@
  */
 
 #include <dt-bindings/media/xilinx-vip.h>
+#include <linux/bitfield.h>
 #include <linux/bitops.h>
 #include <linux/compiler.h>
 #include <linux/clk.h>
@@ -1368,6 +1369,21 @@ static int xsdirx_get_stream_properties(struct xsdirxss_state *state)
 		u8 colorimetry = (payload & XST352_BYTE2_COLORIMETRY_MASK) >>
 			XST352_BYTE2_COLORIMETRY_OFFSET;
 
+		/*
+		 * Bit 7 and 4 of byte 3 form the colorimetry field for HD.
+		 * Checkout SMPTE 292-1:2018 Sec 9.5 for details
+		 */
+		if (mode == XSDIRX_MODE_HD_MASK ||
+		    byte1 == XST352_BYTE1_ST372_DL_3GB) {
+			/* For case when there might be no payload */
+			colorimetry = XST352_BYTE2_COLORIMETRY_BT709;
+
+			if (valid & XSDIRX_ST352_VALID_DS1_MASK) {
+				colorimetry = (FIELD_GET(BIT(23), payload) << 1) |
+					FIELD_GET(BIT(20), payload);
+			}
+		}
+
 		/* Get the EOTF function */
 		switch (eotf) {
 		case XST352_BYTE2_EOTF_SDRTV:
-- 
2.31.1

